README.md
[![Circle CI](https://circleci.com/gh/jshimko/meteor-launchpad/tree/master.svg?style=svg)](https://circleci.com/gh/jshimko/meteor-launchpad/tree/master)
# Meteor Launchpad - Base Docker Image for Meteor Apps

### About
This is a new fork of Meteor-Launchpad that includes 2018 patches.
In particular, this fork includes a patch to the `node-fibers` issue (https://github.com/jshimko/meteor-launchpad/issues/92). Find this project on Dockerhub (https://hub.docker.com/r/nitrolabs/meteor-launchpad/).

### Build

Add the following to a `Dockerfile` in the root of your app:

```Dockerfile
FROM nitrolabs/meteor-launchpad:latest
```

Then you can build the image with:

```sh
docker build -t yourname/app .
```

**Setting up a .dockerignore file**

There are several parts of a Meteor development environment that you don't need to pass into a Docker build because a complete production build happens inside the container.  For example, you don't need to pass in your `node_modules` or the local build files and development database that live in `.meteor/local`.  To avoid copying all of these into the container, here's a recommended starting point for a `.dockerignore` file to be put into the root of your app.  Read more: https://docs.docker.com/engine/reference/builder/#dockerignore-file

```
.git
.meteor/local
node_modules
```

### Run

Now you can run your container with the following command...
(note that the app listens on port 3000 because it is run by a non-root user for [security reasons](https://github.com/nodejs/docker-node/issues/1) and [non-root users can't run processes on port 80](http://stackoverflow.com/questions/16573668/best-practices-when-running-node-js-with-port-80-ubuntu-linode))

```sh
docker run -d \
  -e ROOT_URL=http://example.com \
  -e MONGO_URL=mongodb://url \
  -e MONGO_OPLOG_URL=mongodb://oplog_url \
  -e MAIL_URL=smtp://mail_url.com \
  -p 80:3000 \
  yourname/app
```

#### Delay startup

If you need to force a delay in the startup of the Node process (for example, to wait for a database to be ready), you can set the `STARTUP_DELAY` environment variable to any number of seconds.  For example, to delay starting the app by 10 seconds, you would do this:

```sh
docker run -d \
  -e ROOT_URL=http://example.com \
  -e MONGO_URL=mongodb://url \
  -e STARTUP_DELAY=10 \
  -p 80:3000 \
  yourname/app
```

### Build Options

Meteor Launchpad supports setting custom build options in one of two ways.  You can either create a launchpad.conf config file in the root of your app or you can use [Docker build args](https://docs.docker.com/engine/reference/builder/#arg).  The currently supported options are to install PhantomJS, GraphicsMagick, MongoDB, or any list of `apt-get` dependencies (Meteor Launchpad is built on `debian:jesse`).  

If you choose to install Mongo, you can use it by _not_ supplying a `MONGO_URL` when you run your app container.  The startup script will then start Mongo inside the container and tell your app to use it.  If you _do_ supply a `MONGO_URL`, Mongo will not be started inside the container and the external database will be used instead.

Note that having Mongo in the same container as your app is just for convenience while testing/developing.  In production, you should use a separate Mongo deployment or at least a separate Mongo container.

Here are examples of both methods of setting custom options for your build:

**Option #1 - launchpad.conf**

To use any of them, create a `launchpad.conf` in the root of your app and add any of the following values.

```sh
# launchpad.conf

# Use apt-get to install any additional dependencies
# that you need before your building/running your app
# (default: undefined)
APT_GET_INSTALL="curl git wget"

# Install a custom Node version (default: latest 8.x)
NODE_VERSION=8.9.0

# Installs the latest version of each (default: all false)
INSTALL_MONGO=true
INSTALL_PHANTOMJS=true
INSTALL_GRAPHICSMAGICK=true
```

**Option #2 - Docker Build Args**

If you prefer not to have a config file in your project, your other option is to use the Docker `--build-arg` flag.  When you build your image, you can set any of the same values above as a build arg.

```sh
docker build \
  --build-arg APT_GET_INSTALL="curl git wget" \
  --build-arg INSTALL_MONGO=true \
  --build-arg NODE_VERSION=8.9.0 \
  -t myorg/myapp:latest .
```

## Installing Private NPM Packages

You can provide your [NPM auth token](http://blog.npmjs.org/post/118393368555/deploying-with-npm-private-modules) with the `NPM_TOKEN` build arg.

```sh
docker build --build-arg NPM_TOKEN="<your token>" -t myorg/myapp:latest .
```

## Development Builds

You can optionally avoid downloading Meteor every time when building regularly in development.  Add the following to your Dockerfile instead...

```Dockerfile
FROM nitrolabs/meteor-launchpad:devbuild
```

This isn't recommended for your final production build because it creates a much larger image, but it's a bit of a time saver when you're building often in development.  The first build you run will download/install Meteor and then every subsequent build will be able to skip that step and just build the app.

## Meteor.settings

If you want to include custom settings (as you would via a [settings.json file](https://docs.meteor.com/api/core.html#Meteor-settings)), you need to set the METEOR_SETTINGS environment variable:

```sh
docker run -d \
  -e ROOT_URL=http://example.com \
  -e MONGO_URL=mongodb://url \
  -e MONGO_OPLOG_URL=mongodb://oplog_url \
  -e MAIL_URL=smtp://mail_url.com \
  -e METEOR_SETTINGS="$(cat settings.json)" \
  -p 80:3000 \
  yourname/app
```

## Docker Compose

Add a `docker-compose.yml` to the root of your project with the following content and edit the app image name to match your build name.  Everything else should work as-is.

```yaml
# docker-compose.yml

app:
  image: yourname/app
  ports:
    - "80:3000"
  links:
    - mongo
  environment:
    - ROOT_URL=http://example.com
    - MONGO_URL=mongodb://mongo:27017/meteor

mongo:
  image: mongo:latest
  command: mongod --storageEngine=wiredTiger
```

And then start the app and database containers with...

```sh
docker-compose up -d
```

## Custom Builds of Meteor Launchpad

If you'd like to create a custom build for some reason, you can use the `build.sh` script in the root of the project to run all of the necessary commands.

First, make any changes you want, then to create your custom build:

```sh
# builds as nitrolabs/meteor-launchpad:latest
./build.sh
```

## License

MIT License

Copyright (c) 2017 Jeremy Shimko

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all
copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
SOFTWARE.


HISTORY.md
## v2.4.0
- Fix node-fibers bug introduced with node 8.x
- Fix sporadic issue with gpg keyservers

## v2.3.0

- Updates for Meteor 1.6 support
- Node 8.9.0
- Mongo 3.4.10

## v2.2.2

- Source NODE_VERSION from launchpad.conf if present

## v2.2.1

- Mongo 3.4.9
- add more verbosity to curl calls

## v2.2.0

- add ability to inject an NPM_TOKEN for installation of private packages
- update dockerignore
- add deploy.sh for pushing your built images to a Docker registry

## v2.1.3

- oh, CircleCI... your tag handling sucks sometimes

## v2.1.2

- update GPG key servers to use port 80 (fixes firewall issues for those that can't use non-standard ports)

## v2.1.1

- switch to building the image inside `docker:latest` on CircleCI

## v2.1.0

- use `--server-only` in `meteor build` to skip mobile platform builds
- Mongo 3.4.7
- update CircleCI Docker version

## v2.0.4

- support `APT_GET_INSTALL` from launchpad.conf

## v2.0.3

- add `apt-transport-https` for rare issue (most likely caused by apt source redirects) (#50)
- Node 4.8.4

## v2.0.2

- update CircleCI dependencies

## v2.0.0

- update to CircleCI 2.0 for the automated build/publish
- This update is a fairly big refactor of the Dockerfiles. It removes the shared base Dockerfile and switches to two totally separate builds - the original `devbuild` with every layer aggressively cached (for faster builds in development), and the lean production build with the bare minimum dependencies. Both builds support using `--build-arg` flags to customize what optional dependencies get installed (Mongo, Phantom, etc). The one potentially breaking change here is the `:devbuild` tag no longer supports the usage of the `launchpad.conf` config file because all dependencies are installed/cached before your app code (and config file) is even copied into the container. However, Mongo/Phantom/Graphicsmagick are all set to install by default now, and you can easily override that by running your app build with the following flags (note that this only applies to the devbuild base image - the lean `:latest` image doesn't have those installed unless you specify them with build args):

```sh
# to skip installing Mongo, Phantom, or Graphicsmagick
# when using jshimko/meteor-launchpad:devbuild

docker build \
  --build-arg INSTALL_MONGO=false \
  --build-arg INSTALL_PHANTOMJS=false \
  --build-arg INSTALL_GRAPHICSMAGICK=false \
  -t myorg/myapp:latest .
```

## v1.3.1

- fix issue when no launchpad.conf is found in devbuild

## v1.3.0

- set up all options as Docker build args and update documentation
- remove numactl


## v1.2.0

- allow setting TOOL_NODE_FLAGS for Meteor build at image build time (fixes #41)

Example usage:  
```
docker build --build-arg TOOL_NODE_FLAGS="--max-old-space-size=2048" -t myorg/myapp:latest .
```

## v1.1.8

- replace tar with bsdtar in Meteor installer (fixes #39) thanks @rsercano!


## v1.1.7

- Node 4.8.2
- Mongo 3.4.4


## v1.1.6

- Node 4.8.1
- Mongo 3.4.3


## v1.1.5

- Don't silence Meteor build output
- Update readme with details about creating a custom build
- Mongo 3.4.2
- Make sure Mongo is run by non-root user (if used at all)


## v1.1.4

- Node 4.7.3
- Mongo 3.4.1


## v1.1.2

- Node 4.6.2


## v1.1.1

- If internal Mongo is used, send it to the background (`--fork` required a logfile that would just keep getting larger indefinitely)
- Add default `MONGO_URL` for internal mongod (fixes #22)
- Move `$STARTUP_DELAY` after the mongod startup and add details to the readme
- Send the `meteor build` warning about building as root to /dev/null (since some people have assumed this is an error and the build can take a long time in larger apps).  The warning is irrelevant in this case because all permissions are updated after the build and the app is then run by a non-root user.


## v1.1.0

- update meteor build script to use `METEOR_ALLOW_SUPERUSER` in the build.  Discontinues support of Meteor 1.4.2.  Please update to Meteor 1.4.2.1 if you are currently on 1.4.2
- allow optional install of Graphicsmagick using launchpad.conf - Thanks @un1x86! (#20)
- update gosu to 1.10


## v1.0.0

- Add Docker build versioning to CI build.  Every tag in the form `vX.X.X` will now create and push a new build on CircleCI and update the `:latest` tag. You can now use `jshimko/meteor-launchpad:latest` to always use the latest, or you can pin a specific version with `jshimko/meteor-launchpad:v1.0.0`.  See all available builds here: https://hub.docker.com/r/jshimko/meteor-launchpad/tags/

This is also the last/only version of Meteor Launchpad to support Meteor 1.4.2.  The [issue with using Meteor commands as root](https://github.com/meteor/meteor/issues/7959) introduced in 1.4.2 caused a big problem for this image because it meant the build scripts would _only_ work for Meteor 1.4.2 or later (long story).  [This fix in 1.4.2.1](https://github.com/meteor/meteor/pull/7975/commits/e4acc36f63f98243237b5e6b3e46b083822b95fd) addresses that problem and now allows this image to support any version of Meteor >=1.3 (except 1.4.2 - which you should definitely upgrade if you are using it).


scripts/install-deps.sh
#!/bin/bash

set -e

printf "\n[-] Installing base OS dependencies...\n\n"

# install base dependencies

apt-get update

# ensure we can get an https apt source if redirected
# https://github.com/jshimko/meteor-launchpad/issues/50
apt-get install -y apt-transport-https ca-certificates

if [ -f $APP_SOURCE_DIR/launchpad.conf ]; then
  source <(grep APT_GET_INSTALL $APP_SOURCE_DIR/launchpad.conf)

  if [ "$APT_GET_INSTALL" ]; then
    printf "\n[-] Installing custom apt dependencies...\n\n"
    apt-get install -y $APT_GET_INSTALL
  fi
fi

apt-get install -y --no-install-recommends curl bzip2 bsdtar build-essential python git wget


# install gosu

dpkgArch="$(dpkg --print-architecture | awk -F- '{ print $NF }')"

wget -O /usr/local/bin/gosu "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch"
wget -O /usr/local/bin/gosu.asc "https://github.com/tianon/gosu/releases/download/$GOSU_VERSION/gosu-$dpkgArch.asc"

export GNUPGHOME="$(mktemp -d)"

# Iterate through multiple keyservers, because sometimes they fail
for server in $(shuf -e ha.pool.sks-keyservers.net \
                        hkp://p80.pool.sks-keyservers.net:80 \
                        keyserver.ubuntu.com \
                        hkp://keyserver.ubuntu.com:80 \
                        pgp.mit.edu) ; do \
    gpg --keyserver "$server" --recv-keys B42F6819007F00F88E364FD4036A9C25BF357DD4 && break || : ; \
done && \

gpg --batch --verify /usr/local/bin/gosu.asc /usr/local/bin/gosu

rm -r "$GNUPGHOME" /usr/local/bin/gosu.asc

chmod +x /usr/local/bin/gosu

gosu nobody true

apt-get purge -y --auto-remove wget


scripts/build-meteor.sh
#!/bin/bash

#
# builds a production meteor bundle directory
#
set -e

if [ -f $APP_SOURCE_DIR/launchpad.conf ]; then
  source <(grep TOOL_NODE_FLAGS $APP_SOURCE_DIR/launchpad.conf)
fi

# set up npm auth token if one is provided
if [[ "$NPM_TOKEN" ]]; then
  echo "//registry.npmjs.org/:_authToken=$NPM_TOKEN" >> ~/.npmrc
fi

# Fix permissions warning in Meteor >=1.4.2.1 without breaking
# earlier versions of Meteor with --unsafe-perm or --allow-superuser
# https://github.com/meteor/meteor/issues/7959
export METEOR_ALLOW_SUPERUSER=true

cd $APP_SOURCE_DIR

# Install app deps
printf "\n[-] Running npm install in app directory...\n\n"
meteor npm install

# build the bundle
printf "\n[-] Building Meteor application...\n\n"
mkdir -p $APP_BUNDLE_DIR
meteor build --directory $APP_BUNDLE_DIR --server-only

# run npm install in bundle
printf "\n[-] Running npm install in the server bundle...\n\n"
cd $APP_BUNDLE_DIR/bundle/programs/server/
meteor npm install --production --verbose

# fix fibers bug
# https://github.com/jshimko/meteor-launchpad/issues/92
printf "\n[-] Fixing Fibers PATH bug in node>8.x...\n\n"
cd $APP_BUNDLE_DIR/bundle/programs/server/
npm remove fibers && npm install fibers

# put the entrypoint script in WORKDIR
mv $BUILD_SCRIPTS_DIR/entrypoint.sh $APP_BUNDLE_DIR/bundle/entrypoint.sh

# change ownership of the app to the node user
chown -R node:node $APP_BUNDLE_DIR
